src/util/capabilities.ts

```diff
@@ -0,0 +1,56 @@
+import { MessageConnection, RequestHandler } from 'vscode-jsonrpc';
+import { Protocol } from '../protocol/protocol';
+import { Messages } from '../protocol/messages';
+import { Common, ErrorMessages } from './common';
+import { EventEmitter } from 'events';
+
+/**
+ * Client capabilities handler
+ */
+export class Capabilities {
+
+    private connection: MessageConnection;
+    private emitter: EventEmitter;
+
+    /**
+     * Constructs a new server model handler
+     * @param connection message connection to the RSP
+     * @param emitter event emitter to handle notification events
+     */
+    constructor(connection: MessageConnection, emitter: EventEmitter) {
+        this.connection = connection;
+        this.emitter = emitter;
+        this.listenToClientCapabilities();
+    }
+
+    /**
+     * Subscribe to server creation and deletion events
+     */
+    private listenToClientCapabilities() {
+    let acceptor: (s: string) => void;
+    let rejector: (e: Error) => void;
+    const promise = new Promise<string>((resolve, reject) => {
+        acceptor = (s: string) => {
+            resolve(s);
+        };
+        rejector = (e: Error) => {
+            reject(e);
+        };
+    });
+        const handler1: RequestHandler<Protocol.StringPrompt, String, void> = (p, token) => {
+            this.emitter.emit(Messages.Client.PromptStringRequest.type.method, p, acceptor, rejector);
+            return promise;
```

Comment:
So if I understand this correctly:
 1. you create a promise with just simple resolve and reject callbacks
 2. pass the promise to the handler
 3. the handler fires an event with the resolve & reject callbacks as payload
 4. onStringPrompt gets called
 5. the user calls resolve/reject callback that was passed as the event payload
 6. the promise finishes with whatever was passed to it in onStringPrompt and that is the response to RSP server

Did I decode this right? Anyway, it looks immensely complicated and it doesn't seem too apparent how to actually write the listener for this.

If the event part was skipped and the handler was called directly from onStringPrompt, the listener could then be a simple function.
```typescript
onStringPrompt(listener: (arg: Protocol.StringPrompt) => Promise<string>): void {
  const handler1: RequestHandler<Protocol.StringPrompt, String, void> = p => {
    return listener;
  };
  this.connection.onRequest(Messages.Client.PromptStringRequest.type, handler1);
}
```
Wouldn't that look nicer?